File: bots/stage_1/bot/Field.java
package bot;

import java.util.ArrayList;

public class Field implements FieldInterface {
	public final int COLS = 9, ROWS = 9;

	private int mRoundNr;
	private int mMoveNr;
	private int[][] bigBoard;
	private int[][] smallBoard;

	public Field() {
		bigBoard = new int[ROWS][COLS];
		smallBoard = new int[ROWS / 3][COLS / 3];
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				smallBoard[i][j] = -1;
			}
		}
	}

	// copy constructor
	public Field(Field field) {
		this.mRoundNr = field.mRoundNr;
		this.mMoveNr = field.mMoveNr;

		this.bigBoard = new int[ROWS][COLS];
		for (int i = 0; i < ROWS; i++) {
			for (int j = 0; j < COLS; j++) {
				this.bigBoard[i][j] = field.bigBoard[i][j];
			}
		}
		this.smallBoard = new int[ROWS / 3][COLS / 3];
		for (int i = 0; i < ROWS / 3; i++) {
			for (int j = 0; j < COLS / 3; j++) {
				this.smallBoard[i][j] = field.smallBoard[i][j];
			}
		}
	}

	@Override
	public Winner bigBoardWinner() {
		int i = 0;
		int j = 0;
		int id = -2;
		// horizontal lines
		if (smallBoard[i][j] == smallBoard[i][j + 1]
				&& smallBoard[i][j + 1] == smallBoard[i][j + 2]) {
			if (smallBoard[i][j] > 0) {
				id = smallBoard[i][j];
			}
		}

		else if (smallBoard[i + 1][j] == smallBoard[i + 1][j + 1]
				&& smallBoard[i + 1][j + 1] == smallBoard[i + 1][j + 2]) {
			if (smallBoard[i + 1][j] > 0) {
				id = smallBoard[i + 1][j];
			}
		}

		else if (smallBoard[i + 2][j] == smallBoard[i + 2][j + 1]
				&& smallBoard[i + 2][j + 1] == smallBoard[i + 2][j + 2]) {
			if (smallBoard[i + 2][j] > 0) {
				id = smallBoard[i + 2][j];
			}
		}

		// vertical lines
		else if (smallBoard[i][j] == smallBoard[i + 1][j]
				&& smallBoard[i + 1][j] == smallBoard[i + 2][j]) {
			if (smallBoard[i][j] > 0) {
				id = smallBoard[i][j];
			}
		}

		else if (smallBoard[i][j + 1] == smallBoard[i + 1][j + 1]
				&& smallBoard[i + 1][j + 1] == smallBoard[i + 2][j + 1]) {
			if (smallBoard[i][j + 1] > 0) {
				id = smallBoard[i][j + 1];
			}
		}

		else if (smallBoard[i][j + 2] == smallBoard[i + 1][j + 2]
				&& smallBoard[i + 1][j + 2] == smallBoard[i + 2][j + 2]) {
			if (smallBoard[i][j + 2] > 0) {
				id = smallBoard[i][j + 2];
			}
		}

		// diagonal
		else if (smallBoard[i][j] == smallBoard[i + 1][j + 1]
				&& smallBoard[i + 1][j + 1] == smallBoard[i + 2][j + 2]) {
			if (smallBoard[i][j] > 0) {
				id = smallBoard[i][j];
			}
		}

		else if (smallBoard[i + 2][j] == smallBoard[i + 1][j + 1]
				&& smallBoard[i + 1][j + 1] == smallBoard[i][j + 2]) {
			if (smallBoard[i + 2][j] > 0) {
				id = smallBoard[i + 2][j];
			}
		}

		if (bot.BotParser.mBotId == id)
			return Winner.MYSELF;
		else if (id == BotParser.mOpponentId)
			return Winner.OPPONENT;

		for (int it1 = 0; it1 < 3; it1++) {
			for (int it2 = 0; it2 < 3; it2++) {
				if (smallBoardWinner(it1, it2) == Winner.UNDECIDED)
					return Winner.UNDECIDED;
			}
		}
		return Winner.DRAW;
	}

	@Override
	public Winner smallBoardWinner(int smallBoardRow, int smallBoardCol) {
		int i = smallBoardRow * 3;
		int j = smallBoardCol * 3;
		int id = -2;
		// horizontal lines
		if (bigBoard[i][j] == bigBoard[i][j + 1]
				&& bigBoard[i][j + 1] == bigBoard[i][j + 2]) {
			if (bigBoard[i][j] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i][j];
				id = bigBoard[i][j];
			}
		}

		else if (bigBoard[i + 1][j] == bigBoard[i + 1][j + 1]
				&& bigBoard[i + 1][j + 1] == bigBoard[i + 1][j + 2]) {
			if (bigBoard[i + 1][j] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i +
				// 1][j];
				id = bigBoard[i + 1][j];
			}
		}

		else if (bigBoard[i + 2][j] == bigBoard[i + 2][j + 1]
				&& bigBoard[i + 2][j + 1] == bigBoard[i + 2][j + 2]) {
			if (bigBoard[i + 2][j] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i +
				// 2][j];
				id = bigBoard[i + 2][j];
			}
		}

		// vertical lines
		else if (bigBoard[i][j] == bigBoard[i + 1][j]
				&& bigBoard[i + 1][j] == bigBoard[i + 2][j]) {
			if (bigBoard[i][j] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i][j];
				id = bigBoard[i][j];
			}
		}

		else if (bigBoard[i][j + 1] == bigBoard[i + 1][j + 1]
				&& bigBoard[i + 1][j + 1] == bigBoard[i + 2][j + 1]) {
			if (bigBoard[i][j + 1] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i][j +
				// 1];
				id = bigBoard[i][j + 1];
			}
		}

		else if (bigBoard[i][j + 2] == bigBoard[i + 1][j + 2]
				&& bigBoard[i + 1][j + 2] == bigBoard[i + 2][j + 2]) {
			if (bigBoard[i][j + 2] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i][j +
				// 2];
				id = bigBoard[i][j + 2];
			}
		}

		// diagonal
		else if (bigBoard[i][j] == bigBoard[i + 1][j + 1]
				&& bigBoard[i + 1][j + 1] == bigBoard[i + 2][j + 2]) {
			if (bigBoard[i][j] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i][j];
				id = bigBoard[i][j];
			}
		}

		else if (bigBoard[i + 2][j] == bigBoard[i + 1][j + 1]
				&& bigBoard[i + 1][j + 1] == bigBoard[i][j + 2]) {
			if (bigBoard[i + 2][j] > 0) {
				// smallBoard[smallBoardRow][smallBoardCol] = bigBoard[i +
				// 2][j];
				id = bigBoard[i + 2][j];
			}
		}

		if (bot.BotParser.mBotId == id)
			return Winner.MYSELF;
		else if (id == BotParser.mOpponentId)
			return Winner.OPPONENT;

		for (int it1 = i; it1 < i + 3; it1++) {
			for (int it2 = j; it2 < j + 3; it2++) {
				if (bigBoard[it1][it2] == 0)
					return Winner.UNDECIDED;
			}
		}
		return Winner.DRAW;
	}

	@Override
	public boolean isSmallBoardActive(int smallBoardRow, int smallBoardCol) {
		return smallBoard[smallBoardRow][smallBoardCol] == -1;
	}

	@Override
	public ArrayList<Move> getAvailableMoves() {

		ArrayList<Move> moves = new ArrayList<Move>();
		for (int i = 0; i < ROWS; i++) {
			for (int j = 0; j < COLS; j++) {
				if (isSmallBoardActive(i / 3, j / 3) && bigBoard[i][j] == 0
						&& smallBoardWinner(i / 3, j / 3) == Winner.UNDECIDED) {
					moves.add(new Move(i, j));
				}
			}
		}
		return moves;
	}

	private void updateSmallBoard(int bigBoardRow, int bigBoardCol) {
		// deactivate any active small games
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				if (isSmallBoardActive(i, j)) {
					smallBoard[i][j] = 0;
				}
			}
		}

		int smallBoardRow = bigBoardRow / 3;
		int smallBoardCol = bigBoardCol / 3;

		// decide current small game
		Winner res = smallBoardWinner(smallBoardRow, smallBoardCol);
		switch (res) {
		case MYSELF:
			smallBoard[smallBoardRow][smallBoardCol] = BotParser.mBotId;
			break;
		case OPPONENT:
			smallBoard[smallBoardRow][smallBoardCol] = BotParser.mOpponentId;
			break;
		case DRAW:
		case UNDECIDED:
			smallBoard[smallBoardRow][smallBoardCol] = 0;
			break;
		}

		Winner bigGameWinner = bigBoardWinner();
		if (bigGameWinner != Winner.UNDECIDED) {
			return;
		}

		// finds out the next small board to play in and checks whether it's
		// available
		int nextSmallBoardRow = bigBoardRow - (bigBoardRow / 3) * 3;
		int nextSmallBoardCol = bigBoardCol - (bigBoardCol / 3) * 3;

		if (smallBoardWinner(nextSmallBoardRow, nextSmallBoardCol) == Winner.UNDECIDED) {
			// activate next small game
			smallBoard[nextSmallBoardRow][nextSmallBoardCol] = -1;
		} else {
			// activate all undecided small games
			for (int i = 0; i < 3; i++) {
				for (int j = 0; j < 3; j++) {
					if (smallBoardWinner(i, j) == Winner.UNDECIDED) {
						smallBoard[i][j] = -1;
					}
				}
			}
		}
	}

	@Override
	public boolean placeMove(int bigBoardRow, int bigBoardCol, boolean myself) {
		if (!isSmallBoardActive(bigBoardRow / 3, bigBoardCol / 3))
			return false;
		if (bigBoard[bigBoardRow][bigBoardCol] != 0)
			return false;

		bigBoard[bigBoardRow][bigBoardCol] = (myself) ? BotParser.mBotId
				: BotParser.mOpponentId;

		updateSmallBoard(bigBoardRow, bigBoardCol);

		return true;
	}

	@Override
	public int[][] getBigBoard() {
		return bigBoard;
	}

	@Override
	public int[][] getSmallBoard() {
		return smallBoard;
	}

	@Override
	public int getMoveNumber() {
		return mMoveNr;
	}

	@Override
	public int getRoundNumber() {
		return mRoundNr;
	}

	@Override
	public void parseGameData(String key, String value) {
		if (key.equals("round")) {
			mRoundNr = Integer.parseInt(value);
		} else if (key.equals("move")) {
			mMoveNr = Integer.parseInt(value);
		} else if (key.equals("field")) {
			parseBigBoardFromString(value); /* Parse Field with data */
		} else if (key.equals("macroboard")) {
			parseSmallBoardFromString(value); /* Parse macroboard with data */
		}

	}

	@Override
	public void parseBigBoardFromString(String s) {
		s = s.replace(";", ",");
		String[] r = s.split(",");
		int counter = 0;
		for (int i = 0; i < ROWS; i++) {
			for (int j = 0; j < COLS; j++) {
				bigBoard[i][j] = Integer.parseInt(r[counter]);
				counter++;
			}
		}

	}

	@Override
	public void parseSmallBoardFromString(String s) {
		String[] r = s.split(",");
		int counter = 0;
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				smallBoard[i][j] = Integer.parseInt(r[counter]);
				counter++;
			}
		}
	}
}

File: bots/stage_2/src/Field.cpp
//============================================================================
// Name        : Field.cpp
// Author      : CBTeamName
//============================================================================

#include "Field.h"
#include "GlobalSettings.h"
#include "Utils.h"

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <list>
#include <random>
#include <sstream>
#include <string>
#include <vector>
using namespace std;

/*
 * Default constructor.
 */
Field::Field()
    : gameBoard(ROWS, vector<int>(COLS, 0)),
      smallBoards(ROWS / 3, vector<int>(COLS / 3, -1)) {
    mRoundNr = 1;
    mMoveNr = 1;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            activeSmallBoards.push_back(make_pair(i, j));
        }
    }
}

/*
 * Copy constructor.
 */
Field::Field(const Field& other)
    : mRoundNr(other.mRoundNr),
      mMoveNr(other.mMoveNr),
      gameBoard(other.getGameBoard()),
      smallBoards(other.getSmallBoards()),
      activeSmallBoards(other.activeSmallBoards) {}

/*
 * Determines the winner of the entire game.
 */
Winner Field::gameWinner() const {
    int id = -2;

    // Horizontal lines of small boards.
    for (int i = 0; i <= 2; i++) {
        if (smallBoards[i][0] == smallBoards[i][1] &&
            smallBoards[i][1] == smallBoards[i][2]) {
            if (smallBoards[i][0] > 0) {
                id = smallBoards[i][0];
            }
            break;
        }
    }

    // Vertical lines of small boards.
    for (int j = 0; j <= 2; j++) {
        if (smallBoards[0][j] == smallBoards[1][j] &&
            smallBoards[1][j] == smallBoards[2][j]) {
            if (smallBoards[0][j] > 0) {
                id = smallBoards[0][j];
            }
        }
    }

    // Diagonal lines of small boards.
    if (smallBoards[0][0] == smallBoards[1][1] &&
        smallBoards[1][1] == smallBoards[2][2]) {
        if (smallBoards[0][0] > 0) {
            id = smallBoards[0][0];
        }
    } else if (smallBoards[2][0] == smallBoards[1][1] &&
               smallBoards[1][1] == smallBoards[0][2]) {
        if (smallBoards[2][0] > 0) {
            id = smallBoards[2][0];
        }
    }

    // Verify if bot won the game.
    if (GlobalSettings::getInstance().getBotId() == id) {
        return Winner::MYSELF;
    }
    // Verify if opponent won the game.
    else if (GlobalSettings::getInstance().getOpponentId() == id) {
        return Winner::OPPONENT;
    }

    // Verify if the game is still undecided.
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            if (smallBoardWinner(i, j) == Winner::UNDECIDED) {
                return Winner::UNDECIDED;
            }
        }
    }

    return Winner::DRAW;
}

/*
 * Determines the winner of the small board
 * at position (row, col).
 */
Winner Field::smallBoardWinner(int row, int col) const {
    int botId = GlobalSettings::getInstance().getBotId();
    int opponentId = GlobalSettings::getInstance().getOpponentId();

    if (smallBoards[row][col] == botId) return Winner::MYSELF;
    if (smallBoards[row][col] == opponentId) return Winner::OPPONENT;

    // Determine top-left row and column of the small board.
    int i = row * 3;
    int j = col * 3;
    int id = -2;
    int k;

    // Horizontal lines in the given small board.
    for (k = 0; k <= 2; k++) {
        if (gameBoard[i + k][j] == gameBoard[i + k][j + 1] &&
            gameBoard[i + k][j + 1] == gameBoard[i + k][j + 2]) {
            if (gameBoard[i + k][j] > 0) {
                id = gameBoard[i + k][j];
            }
        }
    }

    // Vertical lines in the given small board.
    for (k = 0; k <= 2; k++) {
        if (gameBoard[i][j + k] == gameBoard[i + 1][j + k] &&
            gameBoard[i + 1][j + k] == gameBoard[i + 2][j + k]) {
            if (gameBoard[i][j + k] > 0) {
                id = gameBoard[i][j + k];
            }
        }
    }

    // Diagonal lines in the given small board.
    if (gameBoard[i][j] == gameBoard[i + 1][j + 1] &&
        gameBoard[i + 1][j + 1] == gameBoard[i + 2][j + 2]) {
        if (gameBoard[i][j] > 0) {
            id = gameBoard[i][j];
        }
    } else if (gameBoard[i + 2][j] == gameBoard[i + 1][j + 1] &&
               gameBoard[i + 1][j + 1] == gameBoard[i][j + 2]) {
        if (gameBoard[i + 2][j] > 0) {
            id = gameBoard[i + 2][j];
        }
    }

    // Verify if bot won this small board.
    if (id == botId) {
        return Winner::MYSELF;
    }
    // Verify if opponent won this small board.
    else if (id == opponentId) {
        return Winner::OPPONENT;
    }

    // Verify if this small board is still undecided.
    for (int it1 = i; it1 < i + 3; it1++) {
        for (int it2 = j; it2 < j + 3; it2++) {
            if (gameBoard[it1][it2] == 0) {
                return Winner::UNDECIDED;
            }
        }
    }

    return Winner::DRAW;
}

/*
 * Checks whether the small board at position
 * (row, col) is active.
 */
inline bool Field::isSmallBoardActive(int row, int col) const {
    return (smallBoards[row][col] == -1);
}

vector<Move> Field::getAvailableMoves() const {
    vector<Move> moves;
    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    for (const pair<int, int>& activeSmallBoard : activeSmallBoards) {
        for (int i = 3 * activeSmallBoard.first;
             i < 3 * (activeSmallBoard.first + 1); i++) {
            for (int j = 3 * activeSmallBoard.second;
                 j < 3 * (activeSmallBoard.second + 1); j++) {
                if (gameBoard[i][j] == 0) {
                    moves.push_back(Move(i, j));
                }
            }
        }
    }
    shuffle(moves.begin(), moves.end(), std::default_random_engine(seed));
    return moves;
}

void Field::updateSmallBoard(int gameBoardRow, int gameBoardCol) {
    // Deactivate any active small board.
    for (const pair<int, int>& activeSmallBoard : activeSmallBoards) {
        smallBoards[activeSmallBoard.first][activeSmallBoard.second] = 0;
    }
    activeSmallBoards.clear();
    int smallBoardRow = gameBoardRow / 3;
    int smallBoardCol = gameBoardCol / 3;

    // Decide current small board.
    Winner smallWinner = smallBoardWinner(smallBoardRow, smallBoardCol);
    switch (smallWinner) {
        case Winner::MYSELF:
            smallBoards[smallBoardRow][smallBoardCol] =
                GlobalSettings::getInstance().getBotId();
            break;
        case Winner::OPPONENT:
            smallBoards[smallBoardRow][smallBoardCol] =
                GlobalSettings::getInstance().getOpponentId();
            break;
        case Winner::DRAW:
        case Winner::UNDECIDED:
            // nothing to do
            break;
    }

    // See if the game has been won.
    Winner bigWinner = gameWinner();
    if (bigWinner != Winner::UNDECIDED) {
        return;
    }

    // Find the next small board that will be activated.
    int nextSmallBoardRow = gameBoardRow - (gameBoardRow / 3) * 3;
    int nextSmallBoardCol = gameBoardCol - (gameBoardCol / 3) * 3;

    if (smallBoardWinner(nextSmallBoardRow, nextSmallBoardCol) ==
        Winner::UNDECIDED) {
        // Activate the next small game.
        smallBoards[nextSmallBoardRow][nextSmallBoardCol] = -1;

        activeSmallBoards.push_back(
            make_pair(nextSmallBoardRow, nextSmallBoardCol));
    } else {
        // Activate all undecided small games.
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (smallBoards[i][j] == 0 &&
                    smallBoardWinner(i, j) == Winner::UNDECIDED) {
                    smallBoards[i][j] = -1;
                    activeSmallBoards.push_back(make_pair(i, j));
                }
            }
        }
    }
}

bool Field::placeMove(int gameBoardRow, int gameBoardCol, bool myself) {
    if (!isSmallBoardActive(gameBoardRow / 3, gameBoardCol / 3)) {
        return false;
    }
    if (gameBoard[gameBoardRow][gameBoardCol] != 0) {
        return false;
    }

    gameBoard[gameBoardRow][gameBoardCol] =
        (myself) ? GlobalSettings::getInstance().getBotId()
                 : GlobalSettings::getInstance().getOpponentId();

    updateSmallBoard(gameBoardRow, gameBoardCol);

    return true;
}

/*
 * Getters
 */
const vector<vector<int> >& Field::getGameBoard() const { return gameBoard; }

const vector<vector<int> >& Field::getSmallBoards() const {
    return smallBoards;
}

int Field::getMoveNumber() const { return mMoveNr; }

int Field::getRoundNumber() const { return mRoundNr; }

/*
 * Communication methods.
 */
void Field::parseGameData(string& key, string& value) {
    if (key == "round") {
        mRoundNr = stringToInt(value);
    } else if (key == "move") {
        mMoveNr = stringToInt(value);
    } else if (key == "field") {
        parseGameBoardFromString(value);
    } else if (key == "macroboard") {
        parseSmallBoardsFromString(value);
    }
}

void Field::parseGameBoardFromString(string& s) {
    stringstream ss(s);
    char delim;
    int temp;
    for (int i = 0; i < ROWS; i++) {
        for (int j = 0; j < COLS; j++) {
            ss >> temp >> delim;
            gameBoard[i][j] = temp;
        }
    }
}

void Field::parseSmallBoardsFromString(string& s) {
    activeSmallBoards.clear();
    stringstream ss(s);
    char delim;
    int temp;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            ss >> temp >> delim;
            smallBoards[i][j] = temp;
            if (temp == -1) activeSmallBoards.push_back(make_pair(i, j));
        }
    }
}

/*
 * Overloaded operators.
 */
ostream& operator<<(ostream& os, const Field& f) {
    os << "Game Board:" << endl;
    for (int i = 0; i < Field::ROWS; i++) {
        for (int j = 0; j < Field::COLS; j++) {
            os << (int)f.gameBoard[i][j] << ' ';
        }
        os << endl;
    }
    os << "Small boards:" << endl;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            os << f.smallBoards[i][j] << ' ';
        }
        os << endl;
    }
    return os;
}

/*
 * Default destructor.
 */
Field::~Field() {
    smallBoards.clear();
    gameBoard.clear();
    activeSmallBoards.clear();
}
